dao-voting/src/main.sw

```diff
@@ -0,0 +1,327 @@
+    /// - `vote_amount` - the amount of votes to cast on the proposal
+    ///
+    /// # Reverts
+    ///
+    /// * When the given proposal id is greater than or equal to proposal_count
+    /// * When the vote amount is 0
+    /// * When the proposal has passed its deadline
+    /// * When the vote amount is greater than the user's deposited balance
+    #[storage(read, write)]fn vote(approve: bool, proposal_id: u64, vote_amount: u64) {
+        validate_id(proposal_id, storage.proposal_count);
+        require(0 < vote_amount, UserError::VoteAmountCannotBeZero);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(height() <= proposal.deadline, ProposalError::ProposalExpired);
+
+        let user = msg_sender().unwrap();
+        let user_balance = storage.balances.get(user);
+
+        require(vote_amount <= user_balance, UserError::InsufficientBalance);
+
+        let mut votes = storage.votes.get((user, proposal_id));
+        if approve {
+            proposal.yes_votes += vote_amount;
+            votes.yes_votes += vote_amount;
+        } else {
+            proposal.no_votes += vote_amount;
+            votes.no_votes += vote_amount;
+        };
+
+        storage.balances.insert(user, user_balance - vote_amount);
+        storage.votes.insert((user, proposal_id), votes);
+        storage.proposals.insert(proposal_id, proposal);
+
+        log(VoteEvent {
+            id: proposal_id, user, vote_amount
+        });
+    }
+
+    /// Execute a given proposal
+    ///
+    /// # Arguments
+    ///
+    /// - `proposal_id` - Identifier used to specifiy a proposal (0 <= proposal_id < proposal_count)
+    ///
+    /// # Reverts
+    ///
+    /// * When the given proposal id is greater than or equal to proposal_count
+    /// * When the proposal is still active and being voted on
+    /// * When the proposal has not met the necessary approval percentage
+    #[storage(read, write)]fn execute(proposal_id: u64) {
+        validate_id(proposal_id, storage.proposal_count);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(!proposal.executed, ProposalError::ProposalExecuted);
+        require(proposal.deadline < height(), ProposalError::ProposalStillActive);
+
+        // TODO figure out how to prevent approval percentage from overflowing
+        // When close to the u64 max
+        // https://github.com/FuelLabs/sway-applications/issues/106
+        let acceptance_percentage = proposal.yes_votes * 100 / (proposal.yes_votes + proposal.no_votes);
+        require(proposal.acceptance_percentage <= acceptance_percentage, ProposalError::InsufficientApprovals);
+
+        asm(call_data: proposal.proposal_transaction.call_data, amount: proposal.proposal_transaction.amount, asset: proposal.proposal_transaction.asset, gas: proposal.proposal_transaction.gas) {
+            call call_data amount asset gas;
+        }
+
+        proposal.executed = true;
```

Comment:
Given that we are calling a contract the proposal update should be above the call to prevent re-entrency. I do not recall if the VM automatically handles that. I know we have a call to check. Let's be safe and move the two lines before the `asm()`